File: built-in plugins dsp/mb-drums-hiphop.cpp
/**
 * MB Hip Hop Kit
 * Category : instrument
 * Type     : drums
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Hard-hitting hip hop drums
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DRUMS_HIPHOP_H
#define MB_DRUMS_HIPHOP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDrumsHiphop : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-drums-hiphop";
    static constexpr const char* PLUGIN_NAME    = "MB Hip Hop Kit";
    static constexpr const char* PLUGIN_TYPE    = "drums";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float thump = 0.8f;  // range [0, 1]
    float volume = 0.85f;  // range [0, 1]
    };

    MbDrumsHiphop() = default;
    ~MbDrumsHiphop() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.thump = std::clamp(params.thump, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Hip Hop Kit
        return input;
    }
};

#endif // MB_DRUMS_HIPHOP_H

File: built-in plugins dsp/mb-drums-hiphop.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDrumsHiphopPlugin: PluginDefinition = {
  id: "mb-drums-hiphop",
  slug: "mb-drums-hiphop",
  name: "MB Hip Hop Kit",
  category: "instrument",
  type: "drums",
  version: "1.0.0",
  description: "Hard-hitting hip hop drums",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 1.0 },
    { type: "noise", detune: 0, gain: 0.4 },
  ],
  envelope: { attack: 0.001, decay: 0.25, sustain: 0.0, release: 0.35 },
  parameters: [
    {
      id: "thump",
      name: "Thump",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.85,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { thump: 0.8, volume: 0.85 },
};

export default MbDrumsHiphopPlugin;
